mindmeld/active_learning/heuristics.py

```diff
@@ -241,6 +258,53 @@ def rank_3d(confidences_3d: List[List[List[float]]]) -> List[int]:
         all_ordered_sample_indices = [MarginSampling.rank_2d(c) for c in confidences_3d]
         return Heuristic.ordered_indices_list_to_final_rank(all_ordered_sample_indices)
 
+    @staticmethod
+    def beam_search_decoder(predictions, top_k=3):
+        output_sequences = [([], 0)]
+
+        # looping through all the predictions
+        for token_probs in predictions:
+            new_sequences = []
+
+            # append new tokens to old sequences and re-score
+            for old_seq, old_score in output_sequences:
+                for char_index, char in enumerate(token_probs):
+                    new_seq = old_seq + [char_index]
+                    # considering log-likelihood for scoring
+                    value = char
+                    if value:
+                        new_score = old_score + math.log(value)
+                    else:
+                        new_score = old_score - math.inf
+                    new_sequences.append((new_seq, new_score))
+
+            # sort all new sequences in the de-creasing order of their score
+            output_sequences = sorted(
+                new_sequences, key=lambda val: val[1], reverse=True
+            )
+
+            # select top-k based on score
+            output_sequences = output_sequences[:top_k]
+
+        return output_sequences
+
+    @staticmethod
+    def rank_entities(entity_confidences: List[List[List[float]]]) -> List[int]:
```

Comment:
A description here explaining the method by which entities are being ranked and where this is normally used would help.